url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/fri/patterns/interpreter/parsergenerator/parsertables/SLRParserTables.java

```diff
@@ -0,0 +1,459 @@
+import fri.patterns.interpreter.parsergenerator.Token;
+import fri.patterns.interpreter.parsergenerator.syntax.*;
+
+/**
+	A table generator, building SLR bottom-up parser tables from a syntax.
+	<p>
+	An artifical START-node gets inserted. Lists of nonterminals and terminals are created.
+	Syntax nodes are created, from which the parser GOTO and PARSE-ACTION tables get built.
+	<p>
+	This class contains dump methods for syntax nodes and FIRST/FOLLOW sets.
+	
+	@author (c) 2000, Fritz Ritzberger
+*/
+
+public class SLRParserTables extends AbstractParserTables
+{
+	protected transient List syntaxNodes = new ArrayList();
+	protected transient FirstSets firstSets;
+	protected transient FollowSets followSets;
+
+
+	/**
+		Appends START symbol and retrieves all symbols. Calls init then.
+		All parser table types run through this constructor.
+	*/
+	public SLRParserTables(Syntax syntax)
+		throws ParserBuildException
+	{
+		this.syntax = addStartSymbol(syntax);	// add START symbol to begin
+		getAllSymbols();
+		init();
+	}
+	
+
+	/** Builds SLR bottom-up parser tables. */
+	protected void init()
+		throws ParserBuildException
+	{
+		syntaxNodes = new SLRSyntaxNode().build(syntax, syntaxNodes, new Hashtable());
+		
+		gotoTable = generateGoto(syntaxNodes);
+
+		Nullable nullAble = new Nullable(syntax, nonterminals);
+		firstSets  = new FirstSets(syntax, nullAble, nonterminals);
+		followSets = new FollowSets(syntax, nullAble, firstSets);
+		
+		parseTable = generateParseAction(syntaxNodes);
+	}
+
+
+	// Looks for top level rules and inserts a START rule pointing to it.
+	// Inserts a default rule pointing to the first rule when none found.
+	private Syntax addStartSymbol(Syntax syntax)
+		throws ParserBuildException
+	{
+		String startSym = null;
+		List startRules = syntax.findStartRules();
+		
+		if (startRules.size() <= 0)	{	// no toplevel rule
+			//throw new ParserBuildException("Grammatik hat keine Start-Regel!");
+			Rule rule = syntax.getRule(0);
+			System.err.println("WARNING: Grammar has no top level rule, taking first rule >"+rule+"<");
+			startSym = rule.getNonterminal();
+		}
+		else
+		if (startRules.size() > 1)	{	// more than one toplevel rules
+			for (int i = 0; i < startRules.size(); i++)	{	// check if all start rules have the same nonterminal on left side
+				Rule r = (Rule) startRules.get(i);
+				String nt = r.getNonterminal();
+
+				if (startSym == null)
+					startSym = nt;
+				else
+				if (startSym.equals(nt) == false)
+					throw new ParserBuildException("Grammar has more than one toplevel rules: "+startRules);
+			}
+		}
+		else	{	// exactly one toplevel rule found
+			startSym = ((Rule) startRules.get(0)).getNonterminal();
+		}
+		
+		Rule start = new Rule("<START>", 1);
+		start.addRightSymbol(startSym);
+		syntax.insertRule(0, start);
+		
+		return syntax;
+	}
+
+
+	/** Returns all symbols (terminals and nonterminals). Builds lists. Called only once on construction. */
+	protected List getAllSymbols()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 99-108, lines 111-130 into separate methods.